### PR TITLE
Changed project structure for pip and unit testing, added travis support

### PR DESCRIPTION
**Pull request by maxwellgerber**
[Build Status](https://travis-ci.org/maxwellgerber/tnote)[Build Status](https://travis-ci.org/maxwellgerber/tnote)[Build Status](https://travis-ci.org/maxwellgerber/tnote)
Complete overhaul of project structure based on what I learned in this blog post:
[https://www.jeffknupp.com/blog/2013/08/16/open-sourcing-a-python-project-the-right-way/](url)
Integrated automatic testing with travis-cl ([travis-ci.org](url))
Added a setup.py for use with pip.
Started a documentation folder.
I ran into a really big problem where I accidentally installed tnote as a module and then would update the tnote.py, but when I imported it into the testing file, it would import the copy of tnote I had installed. Not entirely sure how module importing works, as is evident from my commit history. I'd push to github and travis wouldn't be able to find the module, and then fail. Finally got it to work but it took some time.

If anyone has advice or resources as to how to structure a python project, I'm all ears.

Tomorrow I'm going to try to restructure tnote.py so that we can write tests involving sample databases.
[Build Status](https://travis-ci.org/maxwellgerber/tnote)[Build Status](https://travis-ci.org/maxwellgerber/tnote)[Build Status](https://travis-ci.org/maxwellgerber/tnote)

### CODE REVIEW

**Comment by prodicus** [comment]
I was on the same path. Man you beat me to it :smile: 

Testing it on my machine now.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+language: python
+python:
+  - "3.4"
```
Comment (prodicus) [inline .travis.yml]: Add python2 too, as we did some serious time adding support for it :smile:

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+  - "3.5"
+# command to install dependencies
+install: "pip install -r requirements.txt"
+# command to run tests
+script: py.test
```
Comment (prodicus) [inline .travis.yml]: Never user `py.test`. How different is it from from `unittests`?

File: tnote/test/test_tnote.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+"""Main test class for tnote"""
```
Comment (prodicus) [inline tnote/test/test_tnote.py]: And at last we started with the tests :smile: 

:+1:

**Comment by synlestidae** [comment]
Have you used [http://docs.python-guide.org/en/latest/dev/virtualenvs/](virtualenv) before? It allows you to isolate your pip dependencies... might be helpful!

As for refactoring, what needs to be done is restructuring tnote.py into separate (and more testable) modules e.g. database.py, render.py, etc.

**Comment by prodicus** [comment]
I was on the same path. Man you beat me to it :smile: 

See 

Imgur

Testing it on my machine now.

**Comment by prodicus** [comment]
@maxwellgerber I second @synlestidae on the directory structure. Breaking `tnote.py` into seperate files would make it more readable and maintainable. What' do you think?

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+from __future__ import print_function
```
Comment (prodicus) [inline setup.py]: Awesome!

I landed on jeff's site from the audrey's blog(author for the cookiecutter module).

I checked the setup.py[cookiecutter module](https://github.com/audreyr/cookiecutter) and it looked pretty good.

I highly suggest trying it out if you haven't already.

**Comment by maxwellgerber** [comment]
I agree. Another thing I'd like to see is if we could make the text field of the display scrollable - it doesn't work too well if a note is too long. I'm not sure about using urwid because I don't think it has windows support.

**Comment by synlestidae** [comment]
@maxwellgerber No windows support? No way! Lol. In all seriousness, I am on a Windows machine so I am in favour of cross platform support. Any console libraries that support both? I'm getting on my Google horse to find out...

**Comment by prodicus** [comment]
I am not sure whether `urwid` is supported on windows. If not, we will have to look for an alternative. 

We should start adding test cases too, will save time when adding any new feature

**Comment by synlestidae** [comment]
[http://stackoverflow.com/questions/18177142/input-in-a-python-text-based-gui-tui](Looks like urwid can work in windows under cygwin). Up to you to choose the console library!

**Comment by prodicus** [comment]
Looks like urwid would do fine here if it's supported in cygwin. Any other suggestions guys?

**Comment by maxwellgerber** [comment]
I found [this](https://github.com/dsoprea/Snackwich) but it doesn't really look like a fully-fledged project.
I think urwid might be the way to go on this.

File: .travis.yml (hunk 2 above)
Comment (ismail-s) [inline .travis.yml]: You can write simpler tests with pytest. It also can run unittests, and has an amazing feature called [funcargs/fixtures](http://pytest.org/latest/fixture.html#fixture).

**Comment by ismail-s** [comment]
Urwid is a good choice. Used it once maybe 2 years ago. Resizing the terminal window is handled very well.

**Comment by prodicus** [comment]
@ismail-s How was your experience with it?

**Comment by ismail-s** [comment]
@prodicus Good, better than tkinter where I ended up hacking some things. See https://github.com/ismail-s/ishyChat/blob/master/ishyChat/Client/Views/UrwidView.py. That repo is how I learned a fair bit of programming, so it isn't brilliant (I no longer work on that repo too). Urwid's widgets worked well, and they had detailed docs too. But the auto resizing of everything is awesome and just worked. It's very high level library.

**Comment by ismail-s** [comment]
@prodicus I also forgot to say, but it requires an event loop to run. Although you can use different event loops eg asyncio, twisted, glib, select based.

**Comment by prodicus** [comment]
@ismail-s Will look into that.

**Comment by prodicus** [comment]
Hey @maxwellgerber . I am sorry I couldn't pay attention to this particular PR. Had a lot going on with my classes lately. 

Are we still on with this feature?

**Pull request closed without merging**